src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListener.cs

```diff
@@ -0,0 +1,137 @@
+                }
+
+                break;
+            case WorkspaceChangeKind.ProjectRemoved:
+                RemoveProject(e.OldSolution.GetProject(e.ProjectId));
+                break;
+            case WorkspaceChangeKind.ProjectAdded:
+            case WorkspaceChangeKind.ProjectChanged:
+            case WorkspaceChangeKind.ProjectReloaded:
+            case WorkspaceChangeKind.DocumentAdded:
+            case WorkspaceChangeKind.DocumentRemoved:
+            case WorkspaceChangeKind.DocumentReloaded:
+            case WorkspaceChangeKind.DocumentChanged:
+            case WorkspaceChangeKind.AdditionalDocumentAdded:
+            case WorkspaceChangeKind.AdditionalDocumentRemoved:
+            case WorkspaceChangeKind.AdditionalDocumentReloaded:
+            case WorkspaceChangeKind.AdditionalDocumentChanged:
+            case WorkspaceChangeKind.DocumentInfoChanged:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentAdded:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentChanged:
+                EnqueueUpdate(e.NewSolution.GetProject(e.ProjectId));
+                break;
+            case WorkspaceChangeKind.SolutionCleared:
+            case WorkspaceChangeKind.SolutionRemoved:
+                foreach (var project in e.OldSolution.Projects)
+                {
+                    RemoveProject(project);
+                }
+
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void RemoveProject(Project? project)
+    {
+        if (project?.FilePath is null)
+        {
+            return;
+        }
+
+        TaskDelayScheduler? scheduler;
+        lock (_gate)
+        {
+            if (_workQueues.TryGetValue(project.FilePath, out scheduler))
+            {
+                _workQueues.Remove(project.FilePath);
+            }
+        }
+
+        scheduler?.Dispose();
+    }
+
+    private void EnqueueUpdate(Project? project)
+    {
+        if (_projectRazorJsonFileName is null ||
+            project is not
+            {
+                FilePath: not null,
+                Language: LanguageNames.CSharp
+            })
+        {
+            return;
+        }
+
+        TaskDelayScheduler? scheduler;
+        lock (_gate)
+        {
+            if (!_workQueues.TryGetValue(project.FilePath, out scheduler))
+            {
+                scheduler = new TaskDelayScheduler(s_debounceTime, CancellationToken.None);
+            }
+        }
+
+        scheduler.ScheduleAsyncTask(ct => RazorProjectJsonSerializer.SerializeAsync(project, _projectRazorJsonFileName, ct), CancellationToken.None);
+    }
+
+    public void Dispose()
+    {
+        lock (_gate)
+        {
+            foreach (var kvp in _workQueues)
+            {
+                kvp.Value.Dispose();
+            }
```

Comment:
Add a using directive for `System.Collections.Generic` and you should be able to write this instead. We have a shared `Deconstruct` extension method for `KeyValuePair`.

```suggestion
            foreach (var (_, value) in _workQueues)
            {
                value.Dispose();
            }
```